pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ProducerBuilderFactory.java

```diff
@@ -0,0 +1,158 @@
+ */
+package org.apache.pulsar.functions.instance;
+
+import static org.apache.commons.lang.StringUtils.isEmpty;
+import com.google.common.annotations.VisibleForTesting;
+import java.security.Security;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import lombok.Builder;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.client.api.BatcherBuilder;
+import org.apache.pulsar.client.api.CompressionType;
+import org.apache.pulsar.client.api.CryptoKeyReader;
+import org.apache.pulsar.client.api.HashingScheme;
+import org.apache.pulsar.client.api.MessageRoutingMode;
+import org.apache.pulsar.client.api.ProducerBuilder;
+import org.apache.pulsar.client.api.ProducerCryptoFailureAction;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.common.functions.CryptoConfig;
+import org.apache.pulsar.common.functions.ProducerConfig;
+import org.apache.pulsar.functions.utils.CryptoUtils;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+
+/**
+ * This class is responsible for creating ProducerBuilders with the appropriate configurations to
+ * match the ProducerConfig provided. Producers are created in 2 locations in Pulsar Functions and Connectors
+ * and this class is used to unify the configuration of the producers without duplicating code.
+ */
+@Slf4j
+public class ProducerBuilderFactory {
+
+    private final PulsarClient client;
+    private final ProducerConfig producerConfig;
+    private final Consumer<ProducerBuilder<?>> defaultConfigurer;
+    private final Crypto crypto;
+
+    public ProducerBuilderFactory(PulsarClient client, ProducerConfig producerConfig, ClassLoader functionClassLoader,
+                                  Consumer<ProducerBuilder<?>> defaultConfigurer) {
+        this.client = client;
+        this.producerConfig = producerConfig;
+        this.defaultConfigurer = defaultConfigurer;
+        try {
+            this.crypto = initializeCrypto(functionClassLoader);
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException("Unable to initialize crypto config " + producerConfig.getCryptoConfig(), e);
+        }
+        if (crypto == null) {
+            log.info("crypto key reader is not provided, not enabling end to end encryption");
+        }
+    }
+
+    public <T> ProducerBuilder<T> createProducerBuilder(String topic, Schema<T> schema, String producerName) {
+        ProducerBuilder<T> builder = client.newProducer(schema);
+        if (defaultConfigurer != null) {
+            defaultConfigurer.accept(builder);
+        }
+        builder.blockIfQueueFull(true)
+                .enableBatching(true)
+                .batchingMaxPublishDelay(10, TimeUnit.MILLISECONDS)
+                .hashingScheme(HashingScheme.Murmur3_32Hash) //
+                .messageRoutingMode(MessageRoutingMode.CustomPartition)
+                .messageRouter(FunctionResultRouter.of())
```

Comment:
maybe add `compressionType(CompressionType.LZ4)` here? and if `producerConfig` is not null and provides a `compressionType`, it will override the `compressionType`

Reply:
That would cause a change in behavior. I'd like to avoid changing behavior in the PR in regards to the default compression type. Please see the comments in the code. LZ4 will be set as default for the ContextImpl case. However, in #19470, it removed the default for producers created in PulsarSink.

Reply:
it was removed here: https://github.com/apache/pulsar/pull/19470/files#diff-ec756c3120608bf856e453482b82bf63c3c793d4e9cf751280114aeb27b10687L111